File: AlgoExpert/Binary Search Trees/Find Closest Value In BST.swift
//
//  Find Closest Value In BST.swift
//  AlgoExpert
//
//  Created by Yi-Cheng Lin on 2021/4/11.
//

import Foundation

extension BinarySearchTree {
    func findClosestValueInBST(tree: BST?, target: Int) -> Int {
        guard let tree = tree else { return 0 }
        var closest = tree.value
        search(node: tree, target: target, closest: &closest)
        return closest
    }
    
    private func search(node: BST?, target: Int, closest: inout Int) {
        guard let node = node else { return }
        print(node.value)
        if abs(target - closest) > abs(target - node.value) {
            // 更靠近
            closest = node.value
        }
        
        if target < node.value {
            // 小於所以找左邊
            if let left = node.left {
                search(node: left, target: target, closest: &closest)
            }
        } else if target > node.value {
            // 大於所以找右邊
            if let right = node.right {
                search(node: right, target: target, closest: &closest)
            }
        } else {
            return
        }
    }
    
    
    
    
    
    private func algoSolution(tree: BST?, target: Int, closest: inout Int) -> Int {
        var currentNode = tree
        while currentNode != nil {
            print(currentNode?.value)
            if let node = currentNode {
                let closestDifference = target - closest
                let currentDifference = target - node.value
                
                if closestDifference.magnitude > currentDifference.magnitude {
                    closest = node.value
                }
            }
            
            if let node = currentNode, target < node.value {
                currentNode = node.left
            } else if let node = currentNode, target > node.value {
                currentNode = node.right
            } else {
                break
            }
        }
        
        return closest
    }
}

File: AlgoExpertPy/Binary Search Trees.py/Find Closest Value In BST.py
from BST.BST import BST
def findClosestValueInBst(tree, target):
    
    return findClosestValue(tree, target, float('inf'))

def findClosestValue(node, target, closest):
    currentNode = node
    while currentNode is not None:
        print("current", currentNode.value)
        if abs(currentNode.value - target) < abs(target - closest):
            print("closer")
            closest = currentNode.value
        if target < currentNode.value:
            print("target < currentNode.value(", currentNode.value ,")")
            currentNode = currentNode.left
        elif target > node.value:
            print("target > currentNode.value(", currentNode.value ,")")
            currentNode = currentNode.right
        else: break
        
    return closest


root = BST(10)
root.left = BST(5)
root.left.left = BST(2)
root.left.left.left = BST(1)
root.left.right = BST(5)
root.right = BST(15)
root.right.left = BST(13)
root.right.left.right = BST(14)
root.right.right = BST(22)

print(findClosestValueInBst(root, 12))